### ISSUE TITLE
S3 Connector - Using `+` (plus) symbol causes mismatched request signature

### ISSUE THREAD

**Issue by SeanCheatham** [akka/alpakka#838]
I noticed a couple other closed issues indicating errors when uploading files with either an `@` symbol or a `:` symbol, so I apologize if I should've posted in one of those instead of creating a new one.  There appears to be similar behavior when uploading a file with a `+` (plus) symbol in its name.

This code was run using Alpakka S3 0.17.

Sample Code:
```
val client: S3Client = ???
Source.single(ByteString("Test"))
  .runWith(
    client.multipartUpload("test-bucket", "Special+Character.txt")
  )
```

Produces the following error:
```
The request signature we calculated does not match the signature you provided. Check your key and signing method.
akka.stream.alpakka.s3.S3Exception: The request signature we calculated does not match the signature you provided. Check your key and signing method. (Code: SignatureDoesNotMatch, RequestID: , HostID: )
	at 
akka.stream.alpakka.s3.impl.S3Stream.$anonfun$initiateMultipartUpload$4(S3Stream.scala:216)
```

**Comment by SeanCheatham** [comment]
For reference, [this document](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-keys) lists the characters that are generally safe, characters that require special encoding, and characters that should generally be avoided.

**Comment by tuleism** [comment]
This seems to be the problem that S3 interpretes plus sign as space (see https://forums.aws.amazon.com/thread.jspa?messageID=722673). S3 interpretes `Special+Character.txt` as `Special Character.txt`, while the generated Canonical Request is for `Special+Character.txt`, which leads to mismatched signature error.

I see the way to fix this is to use `Raw-Request-URI` header and manually replace the plus sign by `%2B` (as mentioned in https://github.com/akka/akka-http/issues/451). Any thoughts?

**Comment by SeanCheatham** [comment]
Ah that's interesting indeed!  And a bit disappointing that the issue has existed in S3 for nearly a decade now.

It sounds to me like modifying https://github.com/akka/alpakka/blob/master/s3/src/main/scala/akka/stream/alpakka/s3/impl/HttpRequests.scala#L103 would be a good place to start.

Could possibly modify that method to declare a base HttpRequest:
1. Check to see if the Request Uri contains a `+` symbol
2. If so, stringify it and replace instances of `+` with `%2B`
3. Attach a Raw-Request-URI header to the base Request

I don't *think* doing a naive replacement of `+` with `%2B` would cause issues since the query string should already be encoded, but it may still be a risky approach.

I could try to jump in on a fix for this if you would like, though I'd welcome more feedback on the mentioned changes before I do.  I'd hate to dive into this otherwise excellent alpakka connector only to mess it all up!

**Comment by tuleism** [comment]
Here's my finding:

>If so, stringify it and replace instances of + with %2B

I test this using real S3 key and it works for your example.

>Could possibly modify that method to declare a base HttpRequest

1. Some other actions need to be changed too because they don't go through this `s3Request`. Example: `listBucket`.
2. When `Raw-Request-URI` is included in the base request, you would need to make sure the `Signer` excludes it when creating `CanonicalRequest` at https://github.com/akka/alpakka/blob/master/s3/src/main/scala/akka/stream/alpakka/s3/auth/Signer.scala#L30.

IMO, `Raw-Request-URI` should be included right before the request is sent. Currently the code uses `akka-http`'s `HttpRequest` all over the place. Maybe it's better to have a wrapper around it that only returns the HttpRequest, with or without the custom header, when it is really sent.

**Comment by ennru** [comment]
Fixed in #1738.
